core/store/src/shard_uid_mapping.rs

```diff
@@ -0,0 +1,73 @@
+use crate::{
+    flat::POISONED_LOCK_ERR,
+    io::{Error, Result},
+};
+use near_primitives::shard_layout::ShardUId;
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+
+/// Stores a mapping from ShardUId to ShardUId.
+///
+/// Protected with mutex for concurrent access.
+/// That is for resharding V3 purposes, where we use the mapping strategy for State column.
+#[derive(Clone)]
+pub struct ShardUIdMapping(Arc<Mutex<ShardUIdMappingInner>>);
+
+impl ShardUIdMapping {
+    pub fn new() -> Self {
+        Self(Arc::new(Mutex::new(ShardUIdMappingInner::new())))
+    }
+
+    pub fn map(&self, shard_uid: &ShardUId) -> Option<ShardUId> {
+        self.lock().map(shard_uid)
+    }
+
+    pub fn update(&self, shard_uid: &ShardUId, db_mapped_shard_uid: Option<ShardUId>) -> ShardUId {
+        self.lock().update(shard_uid, db_mapped_shard_uid)
+    }
+
+    fn lock(&self) -> std::sync::MutexGuard<ShardUIdMappingInner> {
+        self.0.lock().expect(POISONED_LOCK_ERR)
+    }
+}
+
+pub fn retrieve_shard_uid_from_db_key(key: &[u8]) -> Result<ShardUId> {
```

Comment:
I would have expected retrieve and replace to be in a shared function, simply something like get new_key or modify_key_for_store which internally handles the logic for replacing the shard_uid?

Also, not a big fan of returning Vec<u8> here, can't do an inplace modification of key by taking &mut key?

Reply:
> I would have expected retrieve and replace to be in a shared function, simply something like get new_key or modify_key_for_store which internally handles the logic for replacing the shard_uid?

Do you suggest making it `ShardUIdMapping::retrieve_shard_uid_from_db_key` static function instead of a standalone function, or something else?

> Also, not a big fan of returning Vec here, can't do an inplace modification of key by taking &mut key?

`key: &[u8]` originally comes from e.g. `Store::get_ser(key: &[u8])` that is used in 200 places.